// @fuel-ts/providers/dist/index.js
"use strict";var Et=Object.create;var v=Object.defineProperty;var Nt=Object.getOwnPropertyDescriptor;var Vt=Object.getOwnPropertyNames;var Ht=Object.getPrototypeOf,vt=Object.prototype.hasOwnProperty;var Wt=(r,t)=>{for(var e in t)v(r,e,{get:t[e],enumerable:!0})},bt=(r,t,e,a)=>{if(t&&typeof t=="object"||typeof t=="function")for(let s of Vt(t))!vt.call(r,s)&&s!==e&&v(r,s,{get:()=>t[s],enumerable:!(a=Nt(t,s))||a.enumerable});return r};var rt=(r,t,e)=>(e=r!=null?Et(Ht(r)):{},bt(t||!r||!r.__esModule?v(e,"default",{value:r,enumerable:!0}):e,r)),zt=r=>bt(v({},"__esModule",{value:!0}),r);var $e={};Wt($e,{ChangeOutputCollisionError:()=>Z,CoinStatus:()=>ot,CreateTransactionRequest:()=>N,MessageStatus:()=>it,NoWitnessAtIndexError:()=>J,NoWitnessByOwnerError:()=>Y,Provider:()=>H,ScriptTransactionRequest:()=>E,TransactionResponse:()=>V,TransactionType:()=>u.TransactionType,buildBlockExplorerUrl:()=>Pe,calculatePriceWithFactor:()=>K,calculateTransactionFee:()=>_,coinQuantityfy:()=>A,getDecodedLogs:()=>Fe,getGasUsedFromReceipts:()=>_t,getReceiptsWithMissingData:()=>yt,inputify:()=>pt,isCoin:()=>B,isMessage:()=>Re,isRawCoin:()=>ut,isRawMessage:()=>Te,normalizeJSON:()=>mt,outputify:()=>lt,returnZeroScript:()=>St,sleep:()=>dt,transactionRequestify:()=>P,withdrawScript:()=>Ae});module.exports=zt($e);var Gt=require("@ethersproject/bytes"),st=require("@fuel-ts/constants"),nt=require("@fuel-ts/math"),A=r=>{var s,n,o,l;let t,e,a;return Array.isArray(r)?(e=r[0],t=(s=r[1])!=null?s:st.NativeAssetId,a=(n=r[2])!=null?n:void 0):(e=r.amount,t=(o=r.assetId)!=null?o:st.NativeAssetId,a=(l=r.max)!=null?l:void 0),{assetId:(0,Gt.hexlify)(t),amount:(0,nt.bn)(e),max:a?(0,nt.bn)(a):void 0}};var c=rt(require("graphql-tag")),ot=(e=>(e.Spent="SPENT",e.Unspent="UNSPENT",e))(ot||{}),it=(e=>(e.Spent="SPENT",e.Unspent="UNSPENT",e))(it||{});var w=c.default`
  fragment transactionFragment on Transaction {
    id
    rawPayload
    gasPrice
    status {
      type: __typename
      ... on SubmittedStatus {
        time
      }
      ... on SuccessStatus {
        block {
          id
        }
        time
        programState {
          returnType
          data
        }
      }
      ... on FailureStatus {
        block {
          id
        }
        time
        reason
      }
    }
  }
`,Tt=c.default`
  fragment receiptFragment on Receipt {
    data
    rawPayload
  }
`,ct=c.default`
  fragment coinFragment on Coin {
    utxoId
    owner
    amount
    assetId
    maturity
    coinStatus: status
    blockCreated
  }
`,Rt=c.default`
  fragment messageFragment on Message {
    amount
    sender
    recipient
    data
    nonce
    messageStatus: status
    daHeight
  }
`,Kt=c.default`
  fragment messageProofFragment on MessageProof {
    proofSet
    proofIndex
    sender
    recipient
    nonce
    amount
    data
    signature
    header {
      id
      daHeight
      transactionsCount
      outputMessagesCount
      transactionsRoot
      outputMessagesRoot
      height
      prevRoot
      time
      applicationHash
    }
  }
`,Ct=c.default`
  fragment balanceFragment on Balance {
    owner
    amount
    assetId
  }
`,Xt=c.default`
  fragment consensusParametersFragment on ConsensusParameters {
    contractMaxSize
    maxInputs
    maxOutputs
    maxWitnesses
    maxGasPerTx
    maxScriptLength
    maxScriptDataLength
    maxStorageSlots
    maxPredicateLength
    maxPredicateDataLength
    gasPriceFactor
    gasPerByte
    maxMessageDataLength
  }
`,W=c.default`
  fragment blockFragment on Block {
    id
    header {
      height
      time
    }
    transactions {
      id
    }
  }
`,Zt=c.default`
  fragment chainInfoFragment on ChainInfo {
    name
    baseChainHeight
    peerCount
    consensusParameters {
      ...consensusParametersFragment
    }
    latestBlock {
      ...blockFragment
    }
  }
  ${Xt}
  ${W}
`,Jt=c.default`
  fragment contractBalanceFragment on ContractBalance {
    contract
    amount
    assetId
  }
`,Yt=c.default`
  query getVersion {
    nodeInfo {
      nodeVersion
    }
  }
`,jt=c.default`
  query getInfo {
    nodeInfo {
      nodeVersion
      minGasPrice
    }
  }
`,te=c.default`
  query getChain {
    chain {
      ...chainInfoFragment
    }
  }
  ${Zt}
`,ee=c.default`
  query getTransaction($transactionId: TransactionId!) {
    transaction(id: $transactionId) {
      ...transactionFragment
    }
  }
  ${w}
`,ae=c.default`
  query getTransactionWithReceipts($transactionId: TransactionId!) {
    transaction(id: $transactionId) {
      ...transactionFragment
      receipts {
        ...receiptFragment
      }
    }
  }
  ${w}
  ${Tt}
`,re=c.default`
  query getTransactions($after: String, $before: String, $first: Int, $last: Int) {
    transactions(after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...transactionFragment
        }
      }
    }
  }
  ${w}
`,se=c.default`
  query getTransactionsByOwner(
    $owner: Address!
    $after: String
    $before: String
    $first: Int
    $last: Int
  ) {
    transactionsByOwner(owner: $owner, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...transactionFragment
        }
      }
    }
  }
  ${w}
`,ne=c.default`
  query getBlock($blockId: BlockId, $blockHeight: U64) {
    block(id: $blockId, height: $blockHeight) {
      ...blockFragment
    }
  }
  ${W}
`,oe=c.default`
  query getBlockWithTransactions($blockId: BlockId, $blockHeight: U64) {
    block(id: $blockId, height: $blockHeight) {
      ...blockFragment
      transactions {
        ...transactionFragment
      }
    }
  }
  ${W}
  ${w}
`,ie=c.default`
  query getBlocks($after: String, $before: String, $first: Int, $last: Int) {
    blocks(after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...blockFragment
        }
      }
    }
  }
  ${W}
`,ce=c.default`
  query getCoin($coinId: UtxoId!) {
    coin(utxoId: $coinId) {
      ...coinFragment
    }
  }
  ${ct}
`,ue=c.default`
  query getCoins(
    $filter: CoinFilterInput!
    $after: String
    $before: String
    $first: Int
    $last: Int
  ) {
    coins(filter: $filter, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...coinFragment
        }
      }
    }
  }
  ${ct}
`,pe=c.default`
  query getResourcesToSpend(
    $owner: Address!
    $queryPerAsset: [SpendQueryElementInput!]!
    $excludedIds: ExcludeInput
  ) {
    resourcesToSpend(owner: $owner, queryPerAsset: $queryPerAsset, excludedIds: $excludedIds) {
      ...coinFragment
      ...messageFragment
    }
  }
  ${ct}
  ${Rt}
`,le=c.default`
  query getContract($contractId: ContractId!) {
    contract(id: $contractId) {
      bytecode
      id
    }
  }
`,ye=c.default`
  query getContractBalance($contract: ContractId!, $asset: AssetId!) {
    contractBalance(contract: $contract, asset: $asset) {
      ...contractBalanceFragment
    }
  }
  ${Jt}
`,ge=c.default`
  query getBalance($owner: Address!, $assetId: AssetId!) {
    balance(owner: $owner, assetId: $assetId) {
      ...balanceFragment
    }
  }
  ${Ct}
`,me=c.default`
  query getBalances(
    $filter: BalanceFilterInput!
    $after: String
    $before: String
    $first: Int
    $last: Int
  ) {
    balances(filter: $filter, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...balanceFragment
        }
      }
    }
  }
  ${Ct}
`,de=c.default`
  query getMessages($owner: Address!, $after: String, $before: String, $first: Int, $last: Int) {
    messages(owner: $owner, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...messageFragment
        }
      }
    }
  }
  ${Rt}
`,Se=c.default`
  query getMessageProof($transactionId: TransactionId!, $messageId: MessageId!) {
    messageProof(transactionId: $transactionId, messageId: $messageId) {
      ...messageProofFragment
    }
  }
  ${Kt}
`,xe=c.default`
  mutation dryRun($encodedTransaction: HexString!, $utxoValidation: Boolean) {
    dryRun(tx: $encodedTransaction, utxoValidation: $utxoValidation) {
      ...receiptFragment
    }
  }
  ${Tt}
`,fe=c.default`
  mutation submit($encodedTransaction: HexString!) {
    submit(tx: $encodedTransaction) {
      id
    }
  }
`,Ie=c.default`
  mutation startSession {
    startSession
  }
`,qe=c.default`
  mutation endSession($sessionId: ID!) {
    endSession(id: $sessionId)
  }
`,he=c.default`
  mutation execute($sessionId: ID!, $op: String!) {
    execute(id: $sessionId, op: $op)
  }
`,be=c.default`
  mutation reset($sessionId: ID!) {
    reset(id: $sessionId)
  }
`,Ge=(r,t,e)=>r();function Bt(r,t=Ge){return{getVersion(e,a){return t(s=>r.request(Yt,e,{...a,...s}),"getVersion","query")},getInfo(e,a){return t(s=>r.request(jt,e,{...a,...s}),"getInfo","query")},getChain(e,a){return t(s=>r.request(te,e,{...a,...s}),"getChain","query")},getTransaction(e,a){return t(s=>r.request(ee,e,{...a,...s}),"getTransaction","query")},getTransactionWithReceipts(e,a){return t(s=>r.request(ae,e,{...a,...s}),"getTransactionWithReceipts","query")},getTransactions(e,a){return t(s=>r.request(re,e,{...a,...s}),"getTransactions","query")},getTransactionsByOwner(e,a){return t(s=>r.request(se,e,{...a,...s}),"getTransactionsByOwner","query")},getBlock(e,a){return t(s=>r.request(ne,e,{...a,...s}),"getBlock","query")},getBlockWithTransactions(e,a){return t(s=>r.request(oe,e,{...a,...s}),"getBlockWithTransactions","query")},getBlocks(e,a){return t(s=>r.request(ie,e,{...a,...s}),"getBlocks","query")},getCoin(e,a){return t(s=>r.request(ce,e,{...a,...s}),"getCoin","query")},getCoins(e,a){return t(s=>r.request(ue,e,{...a,...s}),"getCoins","query")},getResourcesToSpend(e,a){return t(s=>r.request(pe,e,{...a,...s}),"getResourcesToSpend","query")},getContract(e,a){return t(s=>r.request(le,e,{...a,...s}),"getContract","query")},getContractBalance(e,a){return t(s=>r.request(ye,e,{...a,...s}),"getContractBalance","query")},getBalance(e,a){return t(s=>r.request(ge,e,{...a,...s}),"getBalance","query")},getBalances(e,a){return t(s=>r.request(me,e,{...a,...s}),"getBalances","query")},getMessages(e,a){return t(s=>r.request(de,e,{...a,...s}),"getMessages","query")},getMessageProof(e,a){return t(s=>r.request(Se,e,{...a,...s}),"getMessageProof","query")},dryRun(e,a){return t(s=>r.request(xe,e,{...a,...s}),"dryRun","mutation")},submit(e,a){return t(s=>r.request(fe,e,{...a,...s}),"submit","mutation")},startSession(e,a){return t(s=>r.request(Ie,e,{...a,...s}),"startSession","mutation")},endSession(e,a){return t(s=>r.request(qe,e,{...a,...s}),"endSession","mutation")},execute(e,a){return t(s=>r.request(he,e,{...a,...s}),"execute","mutation")},reset(e,a){return t(s=>r.request(be,e,{...a,...s}),"reset","mutation")}}}var S=require("@ethersproject/bytes"),I=require("@fuel-ts/address"),i=require("@fuel-ts/math"),x=require("@fuel-ts/transactions"),Ot=require("graphql-request"),Lt=rt(require("lodash.clonedeep"));var ut=r=>"utxoId"in r,Te=r=>"recipient"in r,B=r=>"id"in r,Re=r=>"recipient"in r;var y=require("@ethersproject/bytes"),z=require("@fuel-ts/constants"),q=require("@fuel-ts/math"),T=require("@fuel-ts/transactions"),pt=r=>{var t,e,a,s,n;switch(r.type){case T.InputType.Coin:{let o=(0,y.arrayify)((t=r.predicate)!=null?t:"0x"),l=(0,y.arrayify)((e=r.predicateData)!=null?e:"0x");return{type:T.InputType.Coin,utxoID:{transactionId:(0,y.hexlify)((0,y.arrayify)(r.id).slice(0,32)),outputIndex:(0,y.arrayify)(r.id)[32]},owner:(0,y.hexlify)(r.owner),amount:(0,q.bn)(r.amount),assetId:(0,y.hexlify)(r.assetId),txPointer:{blockHeight:(0,q.toNumber)((0,y.arrayify)(r.txPointer).slice(0,8)),txIndex:(0,q.toNumber)((0,y.arrayify)(r.txPointer).slice(8,16))},witnessIndex:r.witnessIndex,maturity:(a=r.maturity)!=null?a:0,predicateLength:o.length,predicateDataLength:l.length,predicate:(0,y.hexlify)(o),predicateData:(0,y.hexlify)(l)}}case T.InputType.Contract:return{type:T.InputType.Contract,utxoID:{transactionId:z.ZeroBytes32,outputIndex:0},balanceRoot:z.ZeroBytes32,stateRoot:z.ZeroBytes32,txPointer:{blockHeight:(0,q.toNumber)((0,y.arrayify)(r.txPointer).slice(0,8)),txIndex:(0,q.toNumber)((0,y.arrayify)(r.txPointer).slice(8,16))},contractID:(0,y.hexlify)(r.contractId)};case T.InputType.Message:{let o=(0,y.arrayify)((s=r.predicate)!=null?s:"0x"),l=(0,y.arrayify)((n=r.predicateData)!=null?n:"0x");return{type:T.InputType.Message,sender:(0,y.hexlify)(r.sender),recipient:(0,y.hexlify)(r.recipient),amount:(0,q.bn)(r.amount),nonce:(0,q.bn)(r.nonce),witnessIndex:r.witnessIndex,dataLength:r.data.length,predicateLength:o.length,predicateDataLength:l.length,data:(0,y.hexlify)(r.data),predicate:(0,y.hexlify)(o),predicateData:(0,y.hexlify)(l)}}default:throw new Error("Invalid Input type")}};var h=require("@ethersproject/bytes"),O=require("@fuel-ts/constants"),L=require("@fuel-ts/math"),f=require("@fuel-ts/transactions"),lt=r=>{switch(r.type){case f.OutputType.Coin:return{type:f.OutputType.Coin,to:(0,h.hexlify)(r.to),amount:(0,L.bn)(r.amount),assetId:(0,h.hexlify)(r.assetId)};case f.OutputType.Contract:return{type:f.OutputType.Contract,inputIndex:r.inputIndex,balanceRoot:O.ZeroBytes32,stateRoot:O.ZeroBytes32};case f.OutputType.Message:return{type:f.OutputType.Message,recipient:(0,h.hexlify)(r.recipient),amount:(0,L.bn)(r.amount)};case f.OutputType.Change:return{type:f.OutputType.Change,to:(0,h.hexlify)(r.to),amount:(0,L.bn)(0),assetId:(0,h.hexlify)(r.assetId)};case f.OutputType.Variable:return{type:f.OutputType.Variable,to:O.ZeroBytes32,amount:(0,L.bn)(0),assetId:O.ZeroBytes32};case f.OutputType.ContractCreated:return{type:f.OutputType.ContractCreated,contractId:(0,h.hexlify)(r.contractId),stateRoot:(0,h.hexlify)(r.stateRoot)};default:throw new Error("Invalid Output type")}};var m=require("@ethersproject/bytes"),M=require("@fuel-ts/address"),b=require("@fuel-ts/constants"),U=require("@fuel-ts/math"),u=require("@fuel-ts/transactions");var D=require("@fuel-ts/transactions"),Ce=r=>r.type===D.ReceiptType.Revert&&r.val.toString("hex")===D.FAILED_TRANSFER_TO_ADDRESS_SIGNAL,Be=r=>r.type===D.ReceiptType.Panic&&r.contractId!=="0x0000000000000000000000000000000000000000000000000000000000000000",yt=r=>r.reduce((t,e)=>(Ce(e)&&t.missingOutputVariables.push(e),Be(e)&&t.missingOutputContractIds.push(e),t),{missingOutputVariables:[],missingOutputContractIds:[]});var F=require("@fuel-ts/math"),Q=require("@fuel-ts/transactions"),K=(r,t,e)=>(0,F.bn)(Math.ceil(r.toNumber()/e.toNumber())*t.toNumber()),_t=r=>{let t=r.find(e=>e.type===Q.ReceiptType.ScriptResult);return t&&t.type===Q.ReceiptType.ScriptResult?t.gasUsed:(0,F.bn)(0)},_=({receipts:r,gasPrice:t,margin:e})=>{let a=(0,F.multiply)(_t(r),e||1),s=K(a,t,Q.GAS_PRICE_FACTOR);return{gasUsed:a,fee:s}};var _e="https://fuellabs.github.io/block-explorer-v2",Me=(r,t)=>`${{address:"address",txId:"transaction",blockNumber:"block"}[r]||r}/${t}`,Pe=(r={})=>{let{blockExplorerUrl:t,path:e,providerUrl:a,address:s,txId:n,blockNumber:o}=r,l=t||_e,p=[{key:"address",value:s},{key:"txId",value:n},{key:"blockNumber",value:o}],d=p.filter(G=>!!G.value).map(({key:G,value:at})=>({key:G,value:at})),C=d.length>0;if(d.length>1)throw new Error(`Only one of the following can be passed in to buildBlockExplorerUrl: ${p.map(G=>G.key).join(", ")}`);if(e&&d.length>0){let G=p.map(({key:at})=>at).join(", ");throw new Error(`You cannot pass in a path to buildBlockExplorerUrl along with any of the following: ${G}`)}let Qt=C?Me(d[0].key,d[0].value):"",et=/^\/|\/$/gm,Ft=e?e.replace(et,""):Qt,qt=l.replace(et,""),k=a==null?void 0:a.replace(et,""),ht=k?encodeURIComponent(k):void 0,$t=qt.match(/^https?:\/\//)?"":"https://",Ut=k!=null&&k.match(/^https?:\/\//)?"":"https://";return`${$t}${qt}/${Ft}${ht?`?providerUrl=${Ut}${ht}`:""}`};var Mt=require("@ethersproject/bytes"),Pt=rt(require("lodash.clonedeep"));function gt(r){return Object.keys(r).forEach(t=>{switch(r[t].constructor.name){case"Uint8Array":r[t]=(0,Mt.hexlify)(r[t]);break;case"Array":r[t]=gt(r[t]);break;case"BN":r[t]=r[t].toHex();break;case"Address":r[t]=r[t].toB256();break;case"Object":r[t]=gt(r[t]);break;default:break}}),r}function mt(r){return gt((0,Pt.default)(r))}function dt(r){return new Promise(t=>{setTimeout(()=>{t(!0)},r)})}var $=require("@ethersproject/bytes"),ke=r=>{let t=new Uint8Array(32);return t.set((0,$.arrayify)(r)),t},kt=r=>{let t,e;return Array.isArray(r)?(t=r[0],e=r[1]):(t=r.key,e=r.value),{key:(0,$.hexlify)(t),value:(0,$.hexlify)(ke(e))}};var X=require("@ethersproject/bytes"),At=r=>{let t=(0,X.arrayify)(r);return{data:(0,X.hexlify)(t),dataLength:t.length}};var St={bytes:(0,m.arrayify)("0x24000000"),encodeScriptData:()=>new Uint8Array(0)},Ae={bytes:(0,m.arrayify)("0x5040C0105D44C0064C40001124000000"),encodeScriptData:()=>new Uint8Array(0)},Z=class extends Error{constructor(){super(...arguments);this.name="ChangeOutputCollisionError";this.message='A ChangeOutput with the same "assetId" already exists for a different "to" address'}},J=class extends Error{constructor(e){super();this.index=e;this.name="NoWitnessAtIndexError";this.message=`Witness at index "${e}" was not found`}},Y=class extends Error{constructor(e){super();this.owner=e;this.name="NoWitnessByOwnerError";this.message=`A witness for the given owner "${e}" was not found`}},j=class{constructor({gasPrice:t,gasLimit:e,maturity:a,inputs:s,outputs:n,witnesses:o}={}){this.inputs=[];this.outputs=[];this.witnesses=[];this.gasPrice=(0,U.bn)(t!=null?t:0),this.gasLimit=(0,U.bn)(e!=null?e:0),this.maturity=a!=null?a:0,this.inputs=[...s!=null?s:[]],this.outputs=[...n!=null?n:[]],this.witnesses=[...o!=null?o:[]]}getBaseTransaction(){var s,n,o,l,p,d;let t=(n=(s=this.inputs)==null?void 0:s.map(pt))!=null?n:[],e=(l=(o=this.outputs)==null?void 0:o.map(lt))!=null?l:[],a=(d=(p=this.witnesses)==null?void 0:p.map(At))!=null?d:[];return{gasPrice:this.gasPrice,gasLimit:this.gasLimit,maturity:this.maturity,inputs:t,outputs:e,witnesses:a,inputsCount:t.length,outputsCount:e.length,witnessesCount:a.length}}toTransactionBytes(){return new u.TransactionCoder().encode(this.toTransaction())}pushInput(t){return this.inputs.push(t),this.inputs.length-1}pushOutput(t){return this.outputs.push(t),this.outputs.length-1}createWitness(){return this.witnesses.push("0x"),this.witnesses.length-1}updateWitnessByOwner(t,e){let a=this.getCoinInputWitnessIndexByOwner(t);typeof a=="number"&&this.updateWitness(a,e)}updateWitness(t,e){if(!this.witnesses[t])throw new J(t);this.witnesses[t]=e}getCoinInputs(){return this.inputs.filter(t=>t.type===u.InputType.Coin)}getCoinOutputs(){return this.outputs.filter(t=>t.type===u.OutputType.Coin)}getChangeOutputs(){return this.outputs.filter(t=>t.type===u.OutputType.Change)}getCoinInputWitnessIndexByOwner(t){var a,s,n,o;let e=(0,M.addressify)(t);return(o=(n=(a=this.inputs.find(l=>l.type===u.InputType.Coin&&(0,m.hexlify)(l.owner)===e.toB256()))==null?void 0:a.witnessIndex)!=null?n:(s=this.inputs.find(l=>l.type===u.InputType.Message&&(0,m.hexlify)(l.recipient)===e.toB256()))==null?void 0:s.witnessIndex)!=null?o:null}updateWitnessByCoinInputOwner(t,e){let a=this.getCoinInputWitnessIndexByOwner(t);if(!a)throw new Y((0,M.addressify)(t));this.updateWitness(a,e)}addResource(t){let e=B(t)?t.owner:t.recipient,a=B(t)?t.assetId:b.NativeAssetId,s=B(t)?u.InputType.Coin:u.InputType.Message,n=this.getCoinInputWitnessIndexByOwner(e);typeof n!="number"&&(n=this.createWitness()),this.pushInput(B(t)?{type:s,...t,owner:t.owner.toB256(),witnessIndex:n,txPointer:"0x00000000000000000000000000000000"}:{type:s,...t,sender:t.sender.toB256(),recipient:t.recipient.toB256(),witnessIndex:n,txPointer:"0x00000000000000000000000000000000"});let o=this.getChangeOutputs().find(l=>(0,m.hexlify)(l.assetId)===a);if(o&&(0,m.hexlify)(o.to)!==e.toB256())throw new Z;o||this.pushOutput({type:u.OutputType.Change,to:e.toB256(),assetId:a})}addResources(t){t.forEach(e=>this.addResource(e))}addCoinOutput(t,e,a=b.NativeAssetId){this.pushOutput({type:u.OutputType.Coin,to:(0,M.addressify)(t).toB256(),amount:e,assetId:a})}addCoinOutputs(t,e){e.map(A).forEach(a=>{this.pushOutput({type:u.OutputType.Coin,to:(0,M.addressify)(t).toB256(),amount:a.amount,assetId:a.assetId})})}byteSize(){return this.toTransactionBytes().length}chargeableByteSize(){let t=this.witnesses.reduce((e,a)=>e+(0,m.arrayify)(a).length,0);return(0,U.bn)(this.toTransactionBytes().length-t)}calculateFee(){let t=K(this.gasLimit,this.gasPrice,u.GAS_PRICE_FACTOR);return{assetId:b.NativeAssetId,amount:t.isZero()?(0,U.bn)(1):t}}toJSON(){return mt(this)}},E=class extends j{constructor({script:e,scriptData:a,...s}={}){super(s);this.type=u.TransactionType.Script;this.script=(0,m.arrayify)(e!=null?e:St.bytes),this.scriptData=(0,m.arrayify)(a!=null?a:St.encodeScriptData())}static from(e){return e instanceof this?e:new this(e)}toTransaction(){var s,n;let e=(0,m.arrayify)((s=this.script)!=null?s:"0x"),a=(0,m.arrayify)((n=this.scriptData)!=null?n:"0x");return{type:u.TransactionType.Script,...super.getBaseTransaction(),scriptLength:e.length,scriptDataLength:a.length,receiptsRoot:b.ZeroBytes32,script:(0,m.hexlify)(e),scriptData:(0,m.hexlify)(a)}}getContractInputs(){return this.inputs.filter(e=>e.type===u.InputType.Contract)}getContractOutputs(){return this.outputs.filter(e=>e.type===u.OutputType.Contract)}getVariableOutputs(){return this.outputs.filter(e=>e.type===u.OutputType.Variable)}setScript(e,a){this.script=e.bytes,this.scriptData=e.encodeScriptData(a),this.bytesOffset===void 0&&(this.bytesOffset=this.scriptData.byteLength)}addVariableOutputs(e=1){let a=e;for(;a;)this.pushOutput({type:u.OutputType.Variable}),a-=1;return this.outputs.length-1}addMessageOutputs(e=1){let a=e;for(;a;)this.pushOutput({type:u.OutputType.Message,recipient:"0x0000000000000000000000000000000000000000000000000000000000000000",amount:0}),a-=1;return this.outputs.length-1}addContract(e){let a=(0,M.addressify)(e);if(this.getContractInputs().find(n=>n.contractId===a.toB256()))return;let s=super.pushInput({type:u.InputType.Contract,contractId:a.toB256(),txPointer:"0x00000000000000000000000000000000"});this.pushOutput({type:u.OutputType.Contract,inputIndex:s})}},N=class extends j{constructor({bytecodeWitnessIndex:e,salt:a,storageSlots:s,...n}={}){super(n);this.type=u.TransactionType.Create;this.bytecodeWitnessIndex=e!=null?e:0,this.salt=(0,m.hexlify)(a!=null?a:b.ZeroBytes32),this.storageSlots=[...s!=null?s:[]]}static from(e){return e instanceof this?e:new this(e)}toTransaction(){var n,o;let e=this.getBaseTransaction(),a=this.bytecodeWitnessIndex,s=(o=(n=this.storageSlots)==null?void 0:n.map(kt))!=null?o:[];return{type:u.TransactionType.Create,...e,bytecodeLength:e.witnesses[a].dataLength/4,bytecodeWitnessIndex:a,storageSlotsCount:s.length,salt:this.salt?(0,m.hexlify)(this.salt):b.ZeroBytes32,storageSlots:s}}getContractCreatedOutputs(){return this.outputs.filter(e=>e.type===u.OutputType.ContractCreated)}addContractCreatedOutput(e,a){this.pushOutput({type:u.OutputType.ContractCreated,contractId:e,stateRoot:a})}},P=r=>{if(r instanceof E||r instanceof N)return r;switch(r.type){case u.TransactionType.Script:return E.from(r);case u.TransactionType.Create:return N.from(r);default:throw new Error(`Unknown transaction type: ${r.type}`)}};var xt=require("@ethersproject/bytes"),tt=require("@fuel-ts/math"),R=require("@fuel-ts/transactions");var we=5e3,Oe=1e3,wt=r=>{let t=new R.ReceiptCoder().decode((0,xt.arrayify)(r.rawPayload),0)[0];switch(t.type){case R.ReceiptType.ReturnData:return{...t,data:r.data};case R.ReceiptType.LogData:return{...t,data:r.data};default:return t}},V=class{constructor(t,e){this.gasUsed=(0,tt.bn)(0);this.attempts=0;this.id=t,this.provider=e}async fetch(){let{transaction:t}=await this.provider.operations.getTransactionWithReceipts({transactionId:this.id});return t}decodeTransaction(t){var e;return(e=new R.TransactionCoder().decode((0,xt.arrayify)(t.rawPayload),0))==null?void 0:e[0]}async waitForResult(){var e,a;let t=await this.fetch();switch((e=t==null?void 0:t.status)==null?void 0:e.type){case void 0:case"SubmittedStatus":return this.attempts+=1,await dt(Math.min(Oe*this.attempts,we)),this.waitForResult();case"FailureStatus":{let s=t.receipts.map(wt),{gasUsed:n,fee:o}=_({receipts:s,gasPrice:(0,tt.bn)(t==null?void 0:t.gasPrice)});return this.gasUsed=n,{status:{type:"failure",reason:t.status.reason},receipts:s,transactionId:this.id,blockId:t.status.block.id,time:t.status.time,gasUsed:n,fee:o,transaction:this.decodeTransaction(t)}}case"SuccessStatus":{let s=((a=t.receipts)==null?void 0:a.map(wt))||[],{gasUsed:n,fee:o}=_({receipts:s,gasPrice:(0,tt.bn)(t==null?void 0:t.gasPrice)});return{status:{type:"success",programState:t.status.programState},receipts:s,transactionId:this.id,blockId:t.status.block.id,time:t.status.time,gasUsed:n,fee:o,transaction:this.decodeTransaction(t)}}default:throw new Error("Invalid Transaction status")}}async wait(){let t=await this.waitForResult();if(t.status.type==="failure")throw new Error(`Transaction failed: ${t.status.reason}`);return t}};var Le=10,ft=r=>{let t=new x.ReceiptCoder().decode((0,S.arrayify)(r.rawPayload),0)[0];switch(t.type){case x.ReceiptType.ReturnData:return{...t,data:r.data};case x.ReceiptType.LogData:return{...t,data:r.data};default:return t}},De=r=>{let{name:t,baseChainHeight:e,peerCount:a,consensusParameters:s,latestBlock:n}=r;return{name:t,baseChainHeight:(0,i.bn)(e),peerCount:a,consensusParameters:{contractMaxSize:(0,i.bn)(s.contractMaxSize),maxInputs:(0,i.bn)(s.maxInputs),maxOutputs:(0,i.bn)(s.maxOutputs),maxWitnesses:(0,i.bn)(s.maxWitnesses),maxGasPerTx:(0,i.bn)(s.maxGasPerTx),maxScriptLength:(0,i.bn)(s.maxScriptLength),maxScriptDataLength:(0,i.bn)(s.maxScriptDataLength),maxStorageSlots:(0,i.bn)(s.maxStorageSlots),maxPredicateLength:(0,i.bn)(s.maxPredicateLength),maxPredicateDataLength:(0,i.bn)(s.maxPredicateDataLength),gasPriceFactor:(0,i.bn)(s.gasPriceFactor),gasPerByte:(0,i.bn)(s.gasPerByte),maxMessageDataLength:(0,i.bn)(s.maxMessageDataLength)},latestBlock:{id:n.id,height:(0,i.bn)(n.header.height),time:n.header.time,transactions:n.transactions.map(o=>({id:o.id}))}}},Qe=r=>({minGasPrice:(0,i.bn)(r.minGasPrice),nodeVersion:r.nodeVersion}),H=class{constructor(t,e={}){this.url=t;this.options=e;this.operations=this.createOperations(t,e)}createOperations(t,e={}){this.url=t;let a=new Ot.GraphQLClient(t,e.fetch?{fetch:e.fetch}:void 0);return Bt(a)}connect(t){this.operations=this.createOperations(t)}async getVersion(){let{nodeInfo:{nodeVersion:t}}=await this.operations.getVersion();return t}async getNetwork(){return{name:"fuelv2",chainId:3735928559}}async getBlockNumber(){let{chain:t}=await this.operations.getChain();return(0,i.bn)(t.latestBlock.header.height,10)}async getNodeInfo(){let{nodeInfo:t}=await this.operations.getInfo();return Qe(t)}async getChain(){let{chain:t}=await this.operations.getChain();return De(t)}async sendTransaction(t){let e=P(t);await this.addMissingVariables(e);let a=(0,S.hexlify)(e.toTransactionBytes()),{gasUsed:s,minGasPrice:n}=await this.getTransactionCost(e,0);if((0,i.bn)(s).gt((0,i.bn)(e.gasLimit)))throw new Error(`gasLimit(${e.gasLimit}) is lower than the required (${s})`);if((0,i.bn)(n).gt((0,i.bn)(e.gasPrice)))throw new Error(`gasPrice(${e.gasPrice}) is lower than the required ${n}`);let{submit:{id:o}}=await this.operations.submit({encodedTransaction:a});return new V(o,this)}async call(t,{utxoValidation:e}={}){let a=P(t);await this.addMissingVariables(a);let s=(0,S.hexlify)(a.toTransactionBytes()),{dryRun:n}=await this.operations.dryRun({encodedTransaction:s,utxoValidation:e||!1});return{receipts:n.map(ft)}}async addMissingVariables(t){let e=0,a=0,s=0;if(t.type!==x.TransactionType.Create)do{let n=(0,S.hexlify)(t.toTransactionBytes()),{dryRun:o}=await this.operations.dryRun({encodedTransaction:n,utxoValidation:!1}),l=o.map(ft),{missingOutputVariables:p,missingOutputContractIds:d}=yt(l);if(e=p.length,a=d.length,e===0&&a===0)return;t.addVariableOutputs(e),d.forEach(({contractId:C})=>t.addContract(I.Address.fromString(C))),s+=1}while(s<Le)}async simulate(t){let e=P(t);await this.addMissingVariables(e);let a=(0,S.hexlify)(e.toTransactionBytes()),{dryRun:s}=await this.operations.dryRun({encodedTransaction:a,utxoValidation:!0});return{receipts:s.map(ft)}}async getTransactionCost(t,e=.2){let a=P((0,Lt.default)(t)),{minGasPrice:s}=await this.getNodeInfo(),n=(0,i.max)(a.gasPrice,s),o=1+e;a.gasLimit=x.MAX_GAS_PER_TX,a.gasPrice=(0,i.bn)(0);let{receipts:l}=await this.call(a),{gasUsed:p,fee:d}=_({gasPrice:n,receipts:l,margin:o});return{minGasPrice:s,gasPrice:n,gasUsed:p,fee:d}}async getCoins(t,e,a){return(await this.operations.getCoins({first:10,...a,filter:{owner:t.toB256(),assetId:e&&(0,S.hexlify)(e)}})).coins.edges.map(o=>o.node).map(o=>({id:o.utxoId,assetId:o.assetId,amount:(0,i.bn)(o.amount),owner:I.Address.fromAddressOrString(o.owner),status:o.coinStatus,maturity:(0,i.bn)(o.maturity).toNumber(),blockCreated:(0,i.bn)(o.blockCreated)}))}async getResourcesToSpend(t,e,a){var o,l;let s={messages:((o=a==null?void 0:a.messages)==null?void 0:o.map(p=>(0,S.hexlify)(p)))||[],utxos:((l=a==null?void 0:a.utxos)==null?void 0:l.map(p=>(0,S.hexlify)(p)))||[]};return(await this.operations.getResourcesToSpend({owner:t.toB256(),queryPerAsset:e.map(A).map(({assetId:p,amount:d,max:C})=>({assetId:(0,S.hexlify)(p),amount:d.toString(10),max:C?C.toString(10):void 0})),excludedIds:s})).resourcesToSpend.flat().map(p=>ut(p)?{id:p.utxoId,amount:(0,i.bn)(p.amount),status:p.coinStatus,assetId:p.assetId,owner:I.Address.fromAddressOrString(p.owner),maturity:(0,i.bn)(p.maturity).toNumber(),blockCreated:(0,i.bn)(p.blockCreated)}:{sender:I.Address.fromAddressOrString(p.sender),recipient:I.Address.fromAddressOrString(p.recipient),nonce:(0,i.bn)(p.nonce),amount:(0,i.bn)(p.amount),data:x.InputMessageCoder.decodeData(p.data),status:p.messageStatus,daHeight:(0,i.bn)(p.daHeight)})}async getBlock(t){let e;typeof t=="number"?e={blockHeight:(0,i.bn)(t).toString(10)}:t==="latest"?e={blockHeight:(await this.getBlockNumber()).toString(10)}:e={blockId:(0,i.bn)(t).toString(10)};let{block:a}=await this.operations.getBlock(e);return a?{id:a.id,height:(0,i.bn)(a.header.height),time:a.header.time,transactionIds:a.transactions.map(s=>s.id)}:null}async getBlockWithTransactions(t){let e;typeof t=="number"?e={blockHeight:(0,i.bn)(t).toString(10)}:t==="latest"?e={blockHeight:(await this.getBlockNumber()).toString()}:e={blockId:t};let{block:a}=await this.operations.getBlockWithTransactions(e);return a?{id:a.id,height:(0,i.bn)(a.header.height,10),time:a.header.time,transactionIds:a.transactions.map(s=>s.id),transactions:a.transactions.map(s=>{var n;return(n=new x.TransactionCoder().decode((0,S.arrayify)(s.rawPayload),0))==null?void 0:n[0]})}:null}async getTransaction(t){var a;let{transaction:e}=await this.operations.getTransaction({transactionId:t});return e?(a=new x.TransactionCoder().decode((0,S.arrayify)(e.rawPayload),0))==null?void 0:a[0]:null}async getContract(t){let{contract:e}=await this.operations.getContract({contractId:t});return e||null}async getContractBalance(t,e){let{contractBalance:a}=await this.operations.getContractBalance({contract:t.toB256(),asset:(0,S.hexlify)(e)});return(0,i.bn)(a.amount,10)}async getBalance(t,e){let{balance:a}=await this.operations.getBalance({owner:t.toB256(),assetId:(0,S.hexlify)(e)});return(0,i.bn)(a.amount,10)}async getBalances(t,e){return(await this.operations.getBalances({first:10,...e,filter:{owner:t.toB256()}})).balances.edges.map(n=>n.node).map(n=>({assetId:n.assetId,amount:(0,i.bn)(n.amount)}))}async getMessages(t,e){return(await this.operations.getMessages({first:10,...e,owner:t.toB256()})).messages.edges.map(n=>n.node).map(n=>({sender:I.Address.fromAddressOrString(n.sender),recipient:I.Address.fromAddressOrString(n.recipient),nonce:(0,i.bn)(n.nonce),amount:(0,i.bn)(n.amount),data:x.InputMessageCoder.decodeData(n.data),status:n.messageStatus,daHeight:(0,i.bn)(n.daHeight)}))}async getMessageProof(t,e){let a=await this.operations.getMessageProof({transactionId:t,messageId:e});return a.messageProof?{proofSet:a.messageProof.proofSet,proofIndex:(0,i.bn)(a.messageProof.proofIndex),sender:I.Address.fromAddressOrString(a.messageProof.sender),recipient:I.Address.fromAddressOrString(a.messageProof.recipient),nonce:a.messageProof.nonce,amount:(0,i.bn)(a.messageProof.amount),data:a.messageProof.data,signature:a.messageProof.signature,header:{id:a.messageProof.header.id,daHeight:(0,i.bn)(a.messageProof.header.daHeight),transactionsCount:(0,i.bn)(a.messageProof.header.transactionsCount),outputMessagesCount:(0,i.bn)(a.messageProof.header.outputMessagesCount),transactionsRoot:a.messageProof.header.transactionsRoot,outputMessagesRoot:a.messageProof.header.outputMessagesRoot,height:(0,i.bn)(a.messageProof.header.height),prevRoot:a.messageProof.header.prevRoot,time:a.messageProof.header.time,applicationHash:a.messageProof.header.applicationHash}}:null}};var Dt=require("@fuel-ts/abi-coder"),It=require("@fuel-ts/transactions");function Fe(r,t){return r.reduce((e,a)=>a.type===It.ReceiptType.LogData?e.concat(...t.decodeLog(a.data,a.val1.toNumber(),a.id)):a.type===It.ReceiptType.Log?e.concat(...t.decodeLog(new Dt.U64Coder().encode(a.val0),a.val1.toNumber(),a.id)):e,[])}0&&(module.exports={ChangeOutputCollisionError,CoinStatus,CreateTransactionRequest,MessageStatus,NoWitnessAtIndexError,NoWitnessByOwnerError,Provider,ScriptTransactionRequest,TransactionResponse,TransactionType,buildBlockExplorerUrl,calculatePriceWithFactor,calculateTransactionFee,coinQuantityfy,getDecodedLogs,getGasUsedFromReceipts,getReceiptsWithMissingData,inputify,isCoin,isMessage,isRawCoin,isRawMessage,normalizeJSON,outputify,returnZeroScript,sleep,transactionRequestify,withdrawScript});
